### PR TITLE
Move lz4 decompress to backend executor

### PR DESCRIPTION
**Pull request by Linary**
Change-Id: Idf2e796739fda1dc5ce5ffcfa16bc61b6e1a6281

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 1]
```diff
@@ -230,7 +230,8 @@ private void waitIfBusy() {
             if (this.busyCounter.get() > 0) {
                 synchronized (this) {
                     if (this.busyCounter.get() > 0) {
-                        this.busyCounter.decrementAndGet();
+                        LOG.info("Decrease busy counter: [{}]",
+                                 this.busyCounter.decrementAndGet());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:234]: prefer to define a var

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 2]
```diff
@@ -278,6 +279,16 @@ private boolean isWriteBufferOverflow(Status status) {
                    status.getErrorMsg().contains(expectMsg);
         }
 
+        /**
+         * Maybe useful in the future
+         */
+        private boolean isRPCTimeout(Status status) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:285]: prefer isRpcTimeout

**Review by javeme — COMMENTED** [review]
update the title to "Move lz4 decompress to backend executor"

File: hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java [hunk 3]
```diff
@@ -249,6 +249,22 @@ public static synchronized CoreOptions instance() {
...
+                    "raft.rpc_buf_low_water_mark",
+                    "",
+                    positiveInt(),
+                    10 * 1024 * 1024
+            );
+
+    public static final ConfigOption<Integer> RAFT_RPC_BUF_HIGH_WATER_MARK =
+            new ConfigOption<>(
+                    "raft.rpc_buf_high_water_mark",
+                    "",
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java:263]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java [hunk 4]
```diff
@@ -249,6 +249,22 @@ public static synchronized CoreOptions instance() {
                     60000
             );
 
+    public static final ConfigOption<Integer> RAFT_RPC_BUF_LOW_WATER_MARK =
+            new ConfigOption<>(
+                    "raft.rpc_buf_low_water_mark",
+                    "",
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java:255]: prefer write some description

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java [hunk 5]
```diff
@@ -266,6 +267,9 @@ public void notifyCache(HugeType type, Id id) {
             eventHub = this.params.graphEventHub();
         } else if (type.isSchema()) {
             eventHub = this.params.schemaEventHub();
+            if (id.number() && id.asLong() < 0) {
+                return;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java:271]: add some comment

File: hugegraph-core/src/main/java/com/baidu/hugegraph/util/LZ4Util.java [hunk 6]
```diff
@@ -54,6 +54,10 @@ public static BytesBuffer compress(byte[] bytes, int blockSize,
         } catch (IOException e) {
             throw new BackendException("Failed to compress", e);
         }
+        /*
+         * If need perform reading outside the method,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/util/LZ4Util.java:58]: need to perform

**Review by zhoney — APPROVED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java [hunk 7]
```diff
@@ -107,7 +107,9 @@ private void listenChanges() {
             if ("invalid".equals(args[0])) {
                 HugeType type = (HugeType) args[1];
                 Id id = (Id) args[2];
-                this.arrayCaches.remove(type, id);
+                if (id.number() && id.asLong() > 0) {
+                    this.arrayCaches.remove(type, id);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java:111]: move to remove() method

**Review by javeme — APPROVED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java [hunk 8]
```diff
@@ -322,13 +320,16 @@ public void updateIfNeeded(V schema) {
...
-            if (id.number() && id.asLong() > 0) {
+            if (id.number() && id.asLong() > 0L) {
                 this.set(schema.type(), id, schema);
             }
         }
 
         public V get(HugeType type, Id id) {
-            assert id.number() && id.asLong() > 0 : id;
+            assert id.number();
+            if (id.asLong() <= 0L) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java:330]: define var longId=id.asLong()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java [hunk 9]
```diff
@@ -348,7 +349,10 @@ public V get(HugeType type, Id id) {
         }
 
         public void set(HugeType type, Id id, V value) {
-            assert id.number() && id.asLong() > 0 : id;
+            assert id.number();
+            if (id.asLong() <= 0L) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java:353]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java [hunk 10]
```diff
@@ -373,7 +377,10 @@ public void set(HugeType type, Id id, V value) {
         }
 
         public void remove(HugeType type, Id id) {
-            assert id.number() && id.asLong() > 0 : id;
+            assert id.number();
+            if (id.asLong() <= 0L) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java:381]: ditto

**Review by javeme — COMMENTED** [review]
image
ci failure

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java [hunk 11]
```diff
@@ -348,7 +349,10 @@ public V get(HugeType type, Id id) {
         }
 
         public void set(HugeType type, Id id, V value) {
-            assert id.number() && id.asLong() > 0 : id;
+            assert id.number();
+            if (id.asLong() <= 0L) {
+                return;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java:354]: add `assert false : id`;

**Review by javeme — APPROVED** [review]

**Review by zhoney — APPROVED** [review]

**Pull request merged**